man/man8/tcm_nvdefine.8
.\" Automatically generated by Pod::Man 2.27 (Pod::Simple 3.28)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "tcm_nvdefine 8"
.TH tcm_nvdefine 8 "2018-07-12" "TCM Management" ""
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
tpm_nvdefine
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBtpm_nvdefine [\s-1OPTIONS\s0]\fR
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBtpm_nvdefine\fR defines a new \s-1NVRAM\s0 area at the given index and of given
size. The user has to provide the permissions that control access to
the \s-1NVRAM\s0 area.
.PP
Owner authentication is necessary once the \s-1NVRAM\s0 area 0xFFFFFFFF has been
defined. The owner password may be provided on the command line using
the owner password option.
.PP
The following options are supported:
.IP "\fB\-h, \-\-help\fR" 4
.IX Item "-h, --help"
Display command usage info.
.IP "\fB\-v, \-\-version\fR" 4
.IX Item "-v, --version"
Display command version info.
.IP "\fB\-l, \-\-log [none|error|info|debug]\fR" 4
.IX Item "-l, --log [none|error|info|debug]"
Set logging level.
.IP "\fB\-u, \-\-unicode\fR" 4
.IX Item "-u, --unicode"
Use \s-1TSM UNICODE\s0 encoding for passwords to comply with applications
using \s-1TSM\s0 popup boxes
.IP "\fB\-y, \-\-owner\-well\-known\fR" 4
.IX Item "-y, --owner-well-known"
Use a secret of all zeros (32 bytes of zeros) as the owner's secret.
.IP "\fB\-z, \-\-area\-well\-known\fR" 4
.IX Item "-z, --area-well-known"
Use a secret of all zeros (32 bytes of zeros) as the \s-1NVRAM\s0 area's secret.
.IP "\fB\-o, \-\-pwdo\fR (optional parameter)" 4
.IX Item "-o, --pwdo (optional parameter)"
The owner password.
.Sp
A password may be directly provided for example by using '\-\-pwdo=password' or
\&'\-opassword'. If no password is provided with this option then the program will
prompt the user for the password.
.IP "\fB\-a, \-\-pwda\fR (optional parameter)" 4
.IX Item "-a, --pwda (optional parameter)"
The \s-1NVRAM\s0 area password.
.Sp
A password may be directly provided for example by using '\-\-pwda=password' or
\&'\-apassword'. If no password is provided with this option then the program will
prompt the user for the password.
.IP "\fB\-i, \-\-index\fR" 4
.IX Item "-i, --index"
The index of the \s-1NVRAM\s0 area.
The parameter must either be a decimal number or a hexadecimal number starting with '0x'.
.Sp
To select the \s-1NVRAM\s0 area with index 0x100, the command line parameter should be '\-i 0x100' or
\&'\-\-index 0x100'.
.IP "\fB\-s, \-\-size\fR" 4
.IX Item "-s, --size"
The size of the \s-1NVRAM\s0 area.
The parameter must either be a decimal number or a hexadecimal number starting with '0x'.
.IP "\fB\-p, \-\-permissions\fR" 4
.IX Item "-p, --permissions"
The access permissions associated with the \s-1NVRAM\s0 area.
The parameter must either be a decimal number or a hexadecimal number staring with '0x'. It is
possible to logically 'or' numbers or strings. The following strings are supported:
.RS 4
.IP "\fB\s-1AUTHREAD\s0\fR" 4
.IX Item "AUTHREAD"
Reading requires \s-1NVRAM\s0 area authorization.
.IP "\fB\s-1AUTHWRITE\s0\fR" 4
.IX Item "AUTHWRITE"
Writing requires \s-1NVRAM\s0 area authorization.
.IP "\fB\s-1PPREAD\s0\fR" 4
.IX Item "PPREAD"
Reading requires physical presence.
.IP "\fB\s-1PPWRITE\s0\fR" 4
.IX Item "PPWRITE"
Writing requires physical presence.
.IP "\fB\s-1OWNERREAD\s0\fR" 4
.IX Item "OWNERREAD"
Reading requires owner authorization.
.IP "\fB\s-1OWNERWRITE\s0\fR" 4
.IX Item "OWNERWRITE"
Writing requires owner authorization.
.IP "\fB\s-1GLOBALLOCK\s0\fR" 4
.IX Item "GLOBALLOCK"
A write to index 0 locks the \s-1NVRAM\s0 area until the next TCM_Startup(\s-1ST_CLEAR\s0)
.IP "\fB\s-1READ_STCLEAR\s0\fR" 4
.IX Item "READ_STCLEAR"
A read with size 0 on the same index prevents further reading until the next TCM_Startup(\s-1ST_CLEAR\s0)
.IP "\fB\s-1WRITE_STCLEAR\s0\fR" 4
.IX Item "WRITE_STCLEAR"
A write with size 0 to the same index prevents further writing until the next TCM_Startup(\s-1ST_CLEAR\s0)
.IP "\fB\s-1WRITEDEFINE\s0\fR" 4
.IX Item "WRITEDEFINE"
A write with size 0 to the same index locks the \s-1NVRAM\s0 area permanently
.IP "\fB\s-1WRITEALL\s0\fR" 4
.IX Item "WRITEALL"
The value must be written in a single operation
.RE
.RS 4
.Sp
An example of a permission parameter is:
.Sp
\&\-\-permissions=\*(L"OWNERREAD|OWNERWRITE\*(R"
.RE
.IP "\fB\fR" 4
.IX Item ""
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fBtpm_nvread\fR(8), \fBtpm_nvwrite\fR(8), \fBtpm_nvrelease\fR(8), \fBtpm_nvinfo\fR(8)
.SH "REPORTING BUGS"
.IX Header "REPORTING BUGS"
Report bugs to <trousers\-users@lists.sourceforge.net>


src/tpm_mgmt/tcm_nvdefine.c
/*
 * The Initial Developer of the Original Code is International
 * Business Machines Corporation. Portions created by IBM
 * Corporation are Copyright (C) 2005 International Business
 * Machines Corporation. All Rights Reserved.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the Common Public License as published by
 * IBM Corporation; either version 1 of the License, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * Common Public License for more details.
 *
 * You should have received a copy of the Common Public License
 * along with this program; if not, a copy can be viewed at
 * http://www.opensource.org/licenses/cpl1.0.php.
 */

#include <stdio.h>
#include <errno.h>
#include <limits.h>

#include "tcm_tspi.h"
#include "tcm_utils.h"
#include "tcm_nvcommon.h"

static unsigned int nvindex;
static BOOL nvindex_set;
static unsigned int nvperm;
static unsigned int nvsize;
static const char *ownerpass;
static BOOL ownerWellKnown;
static BOOL askOwnerPass;
static const char *datapass;
static BOOL dataWellKnown;
static BOOL askDataPass;
static int end;
static UINT32 selectedPcrsRead[24];
static UINT32 selectedPcrsWrite[24];
static UINT32 selectedPcrsReadLen = 0;
static UINT32 selectedPcrsWriteLen = 0;
static const char *filename;

TSM_HCONTEXT hContext = 0;


static int parse(const int aOpt, const char *aArg)
{
	switch (aOpt) {
	case 'i':
		if (parseHexOrDecimal(aArg, &nvindex, 0, UINT_MAX,
				      "NVRAM index") != 0)
			return -1;
		nvindex_set = TRUE;
		break;

	case 'p':
		if (!strcmp(aArg, "?")) {
			displayStringsAndValues(permvalues, "");
			end = 1;
			return 0;
		}
		if (parseStringWithValues(aArg, permvalues, &nvperm, UINT_MAX,
				          "NVRAM permission") != 0)
			return -1;
		break;

	case 's':
		if (parseHexOrDecimal(aArg, &nvsize, 0, UINT_MAX,
				      "NVRAM index size") != 0)
			return -1;
		break;

	case 'o':
		ownerpass = aArg;
		if (!ownerpass)
			askOwnerPass = TRUE;
		else
			askOwnerPass = FALSE;
		ownerWellKnown = FALSE;
		break;

	case 'y':
		ownerWellKnown = TRUE;
		ownerpass = NULL;
		askOwnerPass = FALSE;
		break;

	case 'a':
		datapass = aArg;
		if (!datapass)
			askDataPass = TRUE;
		else
			askDataPass = FALSE;
		dataWellKnown = FALSE;
		break;

	case 'z':
		dataWellKnown = TRUE;
		datapass = NULL;
		askDataPass = FALSE;
		break;

	case 'u':
		useUnicode = TRUE;
		break;

	case 'r':
		if (aArg && atoi(aArg) >= 0 && atoi(aArg) < 24) {
			selectedPcrsRead[selectedPcrsReadLen++] = atoi(aArg);
		} else
			return -1;
		break;

	case 'w':
		if (aArg && atoi(aArg) >= 0 && atoi(aArg) < 24) {
			selectedPcrsWrite[selectedPcrsWriteLen++] = atoi(aArg);
		} else
			return -1;
		break;

	case 'f':
		filename = aArg;
		break;

	default:
		return -1;
	}
	return 0;
}

static void help(const char* aCmd)
{
	logCmdHelp(aCmd);
	logUnicodeCmdOption();
	logCmdOption("-y, --owner-well-known",
		     _("Use 32 bytes of zeros (TSM_WELL_KNOWN_SECRET) as the "
		       "TCM owner secret"));
	logCmdOption("-z, --data-well-known",
		     _("Use 32 bytes of zeros (TSM_WELL_KNOWN_SECRET) as the "
		       "NVRAM area's secret"));
	logOwnerPassCmdOption();
	logCmdOption("-a, --pwda",
		     _("NVRAM area password"));
	logNVIndexCmdOption();
	logCmdOption("-s, --size",
		     _("Size of the NVRAM area"));
	logCmdOption("-r, --rpcrs",
		     _("PCRs to seal the NVRAM area to for reading (use multiple times)"));
	logCmdOption("-w, --wpcrs",
		     _("PCRs to seal the NVRAM area to for writing (use multiple times)"));
	logCmdOption("-f, --filename",
		     _("File containing PCR info for the NVRAM area"));

	logCmdOption("-p, --permissions",
		     _("Permissions of the NVRAM area"));
        displayStringsAndValues(permvalues, "                ");
}

void logInvalidPcrInfoFile()
{
	logError(_("Invalid PCR info file. Format is:\n"
		 "[r/w] [PCR IDX] [SHA-1 ascii]\n\nExample:\n"
		 "r 9 00112233445566778899AABBCCDDEEFF00112233"));
}

int
parseNVPermsFile(FILE *f, TSM_HCONTEXT *hContext, TSM_HNVSTORE *nvObject,
		 TSM_HPCRS *hPcrsRead, TSM_HPCRS *hPcrsWrite)
{
	UINT32 pcrSize;
	char rw;
	unsigned int pcr, n;
	char hash_ascii[65], hash_bin[32], save;
	int rc = -1;

	while (!feof(f)) {
		errno = 0;
		n = fscanf(f, "%c %u %s\n", &rw, &pcr, hash_ascii);
		if (n != 3) {
			logInvalidPcrInfoFile();
			goto out;
		} else if (errno != 0) {
			perror("fscanf");
			goto out;
		}

		if (rw != 'r' && rw != 'w') {
			logInvalidPcrInfoFile();
			goto out;
		}

		if (pcr > 15) {
			logError(_("Cannot seal NVRAM area to PCR > 15\n"));
			goto out;
		}

		for (n = 0; n < strlen(hash_ascii); n += 2) {
			save = hash_ascii[n + 2];
			hash_ascii[n + 2] = '\0';
			hash_bin[n/2] = strtoul(&hash_ascii[n], NULL, 16);
			hash_ascii[n + 2] = save;
		}
		pcrSize = n/2;

		if (rw == 'r') {
			if (*hPcrsRead == NULL_HPCRS)
				if (contextCreateObject(*hContext, TSM_OBJECT_TYPE_PCRS,
							TSM_PCRS_STRUCT_INFO_SHORT,
							hPcrsRead) != TSM_SUCCESS)
					goto out;

			if (pcrcompositeSetPcrValue(*hPcrsRead, pcr, pcrSize, (BYTE *)hash_bin)
					!= TSM_SUCCESS)
				goto out;
		} else {
			if (*hPcrsWrite == NULL_HPCRS)
				if (contextCreateObject(*hContext, TSM_OBJECT_TYPE_PCRS,
							TSM_PCRS_STRUCT_INFO_SHORT,
							hPcrsWrite) != TSM_SUCCESS)
					goto out;

			if (pcrcompositeSetPcrValue(*hPcrsWrite, pcr, pcrSize, (BYTE *)hash_bin)
					!= TSM_SUCCESS)
				goto out;
		}
	}

	rc = 0;
out:
	return rc;
}

int main(int argc, char **argv)
{
	TSM_HTCM hTpm;
	TSM_HNVSTORE nvObject;
	TSM_FLAG fNvAttrs;
	TSM_HPOLICY hTpmPolicy, hDataPolicy;
	int iRc = -1;
	BYTE well_known_secret[] = TSM_WELL_KNOWN_SECRET;
	int opswd_len = -1;
	int dpswd_len = -1;
	TSM_HPCRS hPcrsRead = 0, hPcrsWrite = 0;
	struct option hOpts[] = {
		{"index"           , required_argument, NULL, 'i'},
		{"size"            , required_argument, NULL, 's'},
		{"permissions"     , required_argument, NULL, 'p'},
		{"rpcrs"           , required_argument, NULL, 'r'},
		{"wpcrs"           , required_argument, NULL, 'w'},
		{"filename"        , required_argument, NULL, 'f'},
		{"pwdo"            , optional_argument, NULL, 'o'},
		{"pwda"            , optional_argument, NULL, 'a'},
		{"use-unicode"     ,       no_argument, NULL, 'u'},
		{"data-well-known" ,       no_argument, NULL, 'z'},
		{"owner-well-known",       no_argument, NULL, 'y'},
		{NULL              ,       no_argument, NULL, 0},
	};
	TSM_FLAG initFlag = TSM_PCRS_STRUCT_INFO_SHORT;
	UINT32 localityValue = TCM_LOC_ZERO | TCM_LOC_ONE | TCM_LOC_TWO |
			       TCM_LOC_THREE | TCM_LOC_FOUR;

	initIntlSys();

	if (genericOptHandler
		    (argc, argv, "i:s:p:o:a:r:w:f:yzu", hOpts,
		     sizeof(hOpts) / sizeof(struct option), parse, help) != 0)
		goto out;

	if (end) {
		iRc = 0;
		goto out;
	}

	if (!nvindex_set) {
		logError(_("You must provide an index for the NVRAM area.\n"));
		goto out;
	}

	if (nvperm == 0 &&
	    (UINT32)nvindex != TCM_NV_INDEX_LOCK &&
	    (UINT32)nvindex != TCM_NV_INDEX0) {
		logError(_("You must provide permission bits for the NVRAM area.\n"));
		goto out;
	}

	logDebug("permissions = 0x%08x\n", nvperm);

	if (contextCreate(&hContext) != TSM_SUCCESS)
		goto out;

	if (contextConnect(hContext) != TSM_SUCCESS)
		goto out_close;

	if (contextGetTpm(hContext, &hTpm) != TSM_SUCCESS)
		goto out_close;

	fNvAttrs = 0;

	if (contextCreateObject(hContext,
				TSM_OBJECT_TYPE_NV,
				fNvAttrs,
				&nvObject) != TSM_SUCCESS)
		goto out_close;

	if (askOwnerPass) {
		ownerpass = _GETPASSWD(_("Enter owner password: "), &opswd_len,
			FALSE, useUnicode );
		if (!ownerpass) {
			logError(_("Failed to get owner password\n"));
			goto out_close;
		}
	}

	if (ownerpass || ownerWellKnown) {
		if (policyGet(hTpm, &hTpmPolicy) != TSM_SUCCESS)
			goto out_close;
		if (ownerpass) {
			if (opswd_len < 0)
				opswd_len = strlen(ownerpass);
			if (policySetSecret(hTpmPolicy, opswd_len,
					    (BYTE *)ownerpass) != TSM_SUCCESS)
				goto out_close;
		} else {
			if (policySetSecret(hTpmPolicy, TCPA_SM3_256_HASH_LEN,
					    (BYTE *)well_known_secret) != TSM_SUCCESS)
				goto out_close;
		}
	}

	if (askDataPass) {
		datapass = _GETPASSWD(_("Enter NVRAM data password: "), &dpswd_len,
			TRUE, useUnicode );
		if (!datapass) {
			logError(_("Failed to get NVRAM data password\n"));
			goto out_close;
		}
	}

	if (datapass || dataWellKnown) {
		if (contextCreateObject
		    (hContext, TSM_OBJECT_TYPE_POLICY, TSM_POLICY_USAGE,
		     &hDataPolicy) != TSM_SUCCESS)
			goto out_close;

		if (datapass) {
			if (dpswd_len < 0)
				dpswd_len = strlen(datapass);
			if (policySetSecret(hDataPolicy, dpswd_len,
				    (BYTE *)datapass) != TSM_SUCCESS)
				goto out_close;
		} else {
			if (policySetSecret(hDataPolicy, TCPA_SM3_256_HASH_LEN,
				    (BYTE *)well_known_secret) != TSM_SUCCESS)
				goto out_close;
		}

		if (Tspi_Policy_AssignToObject(hDataPolicy, nvObject) !=
		    TSM_SUCCESS)
			goto out_close;
	}

	if (Tspi_SetAttribUint32(nvObject,
				 TSM_TSPATTRIB_NV_INDEX,
				 0,
				 nvindex) != TSM_SUCCESS)
		goto out_close_obj;

	if (Tspi_SetAttribUint32(nvObject,
				 TSM_TSPATTRIB_NV_PERMISSIONS,
				 0,
				 nvperm) != TSM_SUCCESS)
		goto out_close_obj;

	if (Tspi_SetAttribUint32(nvObject,
				 TSM_TSPATTRIB_NV_DATASIZE,
				 0,
				 nvsize) != TSM_SUCCESS)
		goto out_close_obj;

	if (selectedPcrsReadLen) {
		UINT32 pcrSize;
		BYTE *pcrValue;
		UINT32 i;

		for (i = 0; i < selectedPcrsReadLen; i++) {
			if (selectedPcrsRead[i] > 15) {
				logError(_("Cannot seal NVRAM area to PCR > 15\n"));
				goto out_close;
			}
		}

		if (contextCreateObject(hContext, TSM_OBJECT_TYPE_PCRS, initFlag,
					&hPcrsRead) != TSM_SUCCESS)
			goto out_close;

		for (i = 0; i < selectedPcrsReadLen; i++) {
			if (tpmPcrRead(hTpm, selectedPcrsRead[i], &pcrSize, &pcrValue) !=
			    TSM_SUCCESS)
				goto out_close;

			if (pcrcompositeSetPcrValue(hPcrsRead, selectedPcrsRead[i],
						    pcrSize, pcrValue)
					!= TSM_SUCCESS)
				goto out_close;
		}
	}

	if (selectedPcrsWriteLen) {
		UINT32 pcrSize;
		BYTE *pcrValue;
		UINT32 i;

		for (i = 0; i < selectedPcrsWriteLen; i++) {
			if (selectedPcrsWrite[i] > 15) {
				logError(_("Cannot seal NVRAM area to PCR > 15\n"));
				goto out_close;
			}
		}

		if (contextCreateObject(hContext, TSM_OBJECT_TYPE_PCRS, initFlag,
					&hPcrsWrite) != TSM_SUCCESS)
			goto out_close;

		for (i = 0; i < selectedPcrsWriteLen; i++) {
			if (tpmPcrRead(hTpm, selectedPcrsWrite[i], &pcrSize, &pcrValue) !=
			    TSM_SUCCESS)
				goto out_close;

			if (pcrcompositeSetPcrValue(hPcrsWrite, selectedPcrsWrite[i],
						    pcrSize, pcrValue)
					!= TSM_SUCCESS)
				goto out_close;
		}
	}

	if (filename) {
		FILE *f;

		f = fopen(filename, "r");
		if (!f) {
			logError(_("Could not access file '%s'\n"), filename);
			goto out_close_obj;
		}

		if (parseNVPermsFile(f, &hContext, &nvObject, &hPcrsRead, &hPcrsWrite)
		    != TSM_SUCCESS)
			goto out_close_obj;
	}

	if (hPcrsRead)
		if (pcrcompositeSetPcrLocality(hPcrsRead, localityValue) != TSM_SUCCESS)
			goto out_close;

	if (hPcrsWrite)
		if (pcrcompositeSetPcrLocality(hPcrsWrite, localityValue) != TSM_SUCCESS)
			goto out_close;

	if (NVDefineSpace(nvObject, hPcrsRead, hPcrsWrite) != TSM_SUCCESS)
		goto out_close;

	logMsg(_("Successfully created NVRAM area at index 0x%x (%u).\n"),
	       nvindex, nvindex);

	iRc = 0;

	goto out_close;

      out_close_obj:
	contextCloseObject(hContext, nvObject);

      out_close:
	contextClose(hContext);

      out:
	return iRc;
}
